File: backend/app/services/admin_service.py
"""
Admin service matching Java AdminService.java.
Admin-only operations for user and company management.
"""
from typing import Optional
from sqlalchemy.orm import Session

from app.exceptions.custom_exceptions import ResourceNotFoundException, DuplicateResourceException
from app.models.user import Role
from app.models.company import Company
from app.repositories.user_repository import UserRepository
from app.repositories.company_repository import CompanyRepository
from app.schemas.admin import (
    UserSummaryResponse,
    PlatformStatsResponse,
    CompanyResponse,
    PageResponse,
    CreateCompanyRequest
)


class AdminService:
    """
    Admin-only operations. Kept intentionally small for Month 1.
    Matches Java AdminService.java.
    """
    
    def __init__(self, db: Session):
        self.db = db
        self.user_repo = UserRepository(db)
        self.company_repo = CompanyRepository(db)
    
    def list_users(
        self,
        role: Optional[Role],
        page: int,
        size: int
    ) -> PageResponse:
        """
        List users with optional role filter and pagination.
        Matches Java: @Transactional(readOnly = true) public Page<UserSummaryResponse> listUsers(...)
        """
        if role:
            users, total = self.user_repo.find_by_role(role, page, size)
        else:
            users, total = self.user_repo.find_all(page, size)
        
        content = [
            UserSummaryResponse(
                id=u.id,
                name=u.name,
                email=u.email,
                role=u.role,
                enabled=u.enabled,
                created_at=u.created_at
            )
            for u in users
        ]
        
        return PageResponse.create(content, page, size, total)
    
    def get_stats(self) -> PlatformStatsResponse:
        """
        Get platform statistics.
        Matches Java: @Transactional(readOnly = true) public PlatformStatsResponse getStats()
        """
        return PlatformStatsResponse(
            total_users=self.user_repo.count_all(),
            total_candidates=self.user_repo.count_by_role(Role.CANDIDATE),
            total_recruiters=self.user_repo.count_by_role(Role.RECRUITER),
            total_admins=self.user_repo.count_by_role(Role.ADMIN),
            total_companies=self.company_repo.count_all()
        )
    
    def set_user_enabled(self, user_id: int, enabled: bool) -> UserSummaryResponse:
        """
        Enable or disable a user account.
        Matches Java: @Transactional public UserSummaryResponse setUserEnabled(...)
        """
        user = self.user_repo.find_by_id(user_id)
        if not user:
            raise ResourceNotFoundException(f"User not found: {user_id}")
        
        user.enabled = enabled
        user = self.user_repo.save(user)
        self.db.commit()
        
        return UserSummaryResponse(
            id=user.id,
            name=user.name,
            email=user.email,
            role=user.role,
            enabled=user.enabled,
            created_at=user.created_at
        )
    
    def create_company(self, request: CreateCompanyRequest) -> CompanyResponse:
        """
        Create a new company.
        Matches Java: @Transactional public Company createCompany(...)
        """
        name = request.name.strip()
        
        if self.company_repo.exists_by_name_ignore_case(name):
            raise DuplicateResourceException(f"A company named '{name}' already exists.")
        
        company = Company(
            name=name,
            description=request.description,
            website=request.website,
            location=request.location
        )
        company = self.company_repo.save(company)
        self.db.commit()
        
        return self._company_to_response(company)
    
    def list_companies(self, page: int, size: int) -> PageResponse:
        """
        List all companies with pagination.
        Matches Java: @Transactional(readOnly = true) public Page<Company> listCompanies(...)
        """
        companies, total = self.company_repo.find_all(page, size)
        
        content = [self._company_to_response(c) for c in companies]
        
        return PageResponse.create(content, page, size, total)
    
    def _company_to_response(self, company: Company) -> CompanyResponse:
        """Convert Company entity to CompanyResponse DTO."""
        return CompanyResponse(
            id=company.id,
            name=company.name,
            description=company.description,
            website=company.website,
            location=company.location,
            created_at=company.created_at
        )

File: backend/src/main/java/com/codeforge/service/AdminService.java
package com.codeforge.service;

import com.codeforge.dto.request.CreateCompanyRequest;
import com.codeforge.dto.response.CompanyResponse;
import com.codeforge.dto.response.PageResponse;
import com.codeforge.dto.response.PlatformStatsResponse;
import com.codeforge.dto.response.UserSummaryResponse;
import com.codeforge.entity.Company;
import com.codeforge.entity.User;
import com.codeforge.entity.User.UserRole;
import com.codeforge.exception.DuplicateResourceException;
import com.codeforge.exception.ResourceNotFoundException;
import com.codeforge.repository.CandidateRepository;
import com.codeforge.repository.CompanyRepository;
import com.codeforge.repository.ProjectRepository;
import com.codeforge.repository.RecruiterRepository;
import com.codeforge.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.format.DateTimeFormatter;

/**
 * Service for admin operations
 */
@Service
@RequiredArgsConstructor
public class AdminService {
    
    private final UserRepository userRepository;
    private final CandidateRepository candidateRepository;
    private final RecruiterRepository recruiterRepository;
    private final CompanyRepository companyRepository;
    private final ProjectRepository projectRepository;
    
    /**
     * Get platform statistics
     */
    public PlatformStatsResponse getPlatformStats() {
        long totalUsers = userRepository.count();
        long totalCandidates = userRepository.countByRole(UserRole.CANDIDATE);
        long totalRecruiters = userRepository.countByRole(UserRole.RECRUITER);
        long totalCompanies = companyRepository.count();
        long totalProjects = projectRepository.count();
        
        long enabledUsers = userRepository.findAll().stream()
            .filter(User::getEnabled)
            .count();
        long disabledUsers = totalUsers - enabledUsers;
        
        return PlatformStatsResponse.builder()
            .totalUsers(totalUsers)
            .totalCandidates(totalCandidates)
            .totalRecruiters(totalRecruiters)
            .totalCompanies(totalCompanies)
            .totalProjects(totalProjects)
            .enabledUsers(enabledUsers)
            .disabledUsers(disabledUsers)
            .build();
    }
    
    /**
     * List users with pagination and optional role filter
     */
    public PageResponse<UserSummaryResponse> listUsers(UserRole role, Pageable pageable) {
        Page<User> page;
        
        if (role != null) {
            page = userRepository.findByRole(role, pageable);
        } else {
            page = userRepository.findAll(pageable);
        }
        
        DateTimeFormatter formatter = DateTimeFormatter.ISO_DATE_TIME;
        
        return PageResponse.<UserSummaryResponse>builder()
            .content(page.getContent().stream()
                .map(user -> UserSummaryResponse.builder()
                    .id(user.getId())
                    .name(user.getName())
                    .email(user.getEmail())
                    .role(user.getRole().name())
                    .enabled(user.getEnabled())
                    .createdAt(user.getCreatedAt().format(formatter))
                    .build())
                .toList())
            .page(page.getNumber())
            .size(page.getSize())
            .totalPages(page.getTotalPages())
            .totalElements(page.getTotalElements())
            .last(page.isLast())
            .build();
    }
    
    /**
     * Set user enabled/disabled status
     */
    @Transactional
    public UserSummaryResponse setUserEnabled(Long userId, boolean enabled) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> ResourceNotFoundException.notFound("User", "id", userId));
        
        user.setEnabled(enabled);
        user = userRepository.save(user);
        
        DateTimeFormatter formatter = DateTimeFormatter.ISO_DATE_TIME;
        
        return UserSummaryResponse.builder()
            .id(user.getId())
            .name(user.getName())
            .email(user.getEmail())
            .role(user.getRole().name())
            .enabled(user.getEnabled())
            .createdAt(user.getCreatedAt().format(formatter))
            .build();
    }
    
    /**
     * Create a new company
     */
    @Transactional
    public CompanyResponse createCompany(CreateCompanyRequest request) {
        if (companyRepository.existsByName(request.getName())) {
            throw DuplicateResourceException.alreadyExists("Company", "name", request.getName());
        }
        
        Company company = Company.builder()
            .name(request.getName())
            .description(request.getDescription())
            .website(request.getWebsite())
            .logoUrl(request.getLogoUrl())
            .build();
        
        company = companyRepository.save(company);
        
        return mapToCompanyResponse(company);
    }
    
    /**
     * List companies with pagination
     */
    public PageResponse<CompanyResponse> listCompanies(Pageable pageable) {
        Page<Company> page = companyRepository.findAll(pageable);
        
        DateTimeFormatter formatter = DateTimeFormatter.ISO_DATE_TIME;
        
        return PageResponse.<CompanyResponse>builder()
            .content(page.getContent().stream()
                .map(this::mapToCompanyResponse)
                .toList())
            .page(page.getNumber())
            .size(page.getSize())
            .totalPages(page.getTotalPages())
            .totalElements(page.getTotalElements())
            .last(page.isLast())
            .build();
    }
    
    private CompanyResponse mapToCompanyResponse(Company company) {
        DateTimeFormatter formatter = DateTimeFormatter.ISO_DATE_TIME;
        
        return CompanyResponse.builder()
            .id(company.getId())
            .name(company.getName())
            .description(company.getDescription())
            .website(company.getWebsite())
            .logoUrl(company.getLogoUrl())
            .createdAt(company.getCreatedAt().format(formatter))
            .build();
    }
}
